### ISSUE TITLE
Upgrade NATS to v2.x

### ISSUE THREAD

**Issue by ryanhallware** [cloudfoundry/nats-release#26]
In issue #25 one of the maintainers of NATS.io [suggested](https://github.com/cloudfoundry/nats-release/pull/22#issuecomment-692822270):

> Hi, looking at issues like #25 I would recommend upgrading to a NATS v2 based setup since it includes many fixes to Slow Consumer scenarios thanks to the reworked clustering protocol. It would be a drop in replacement for the clients, but during the upgrade there would be a reconnection between the servers. Let me know if you need more info, happy to help with this.

The last time we upgraded NATS was in February 2020 when we upgraded from v1.1.0 to v1.4.1. Since then there have been no new releases of NATS v1.x. Given this, in order to stay up-to-date, we should upgrade NATS to a v2.x release. 

When doing this, let's be sure to check how this behaves during our upgrades including connections between the NATS server instances, and connections between the servers and our main clients in Cloud Foundry (route-emitter, gorouter, etc).

**Comment by wallyqs** [comment]
Hi! Hearing from the CF community seems that besides the performance upgrades due to the clustering protocol rewrite, there are a few other things that would improve the behavior and setup of the system if moving to NATS v2:

- Mixed TLS and non/TLS option `allow_non_tls: true`.
  https://github.com/nats-io/nats-server/pull/1458

  I believe that currently in CloudFoundry there are co-located NATS processes in the same VM, one with TLS and another one without TLS, that are clustered but with route advertising disabled to avoid issues like: https://github.com/cloudfoundry/nats-release/issues/25

  Using that mode would simplify the setup and then no need to disable cluster advertise to the clients so that GoRouters can reconnect faster to the discovered NATS servers.

- Adopting lame duck mode to gracefully shutdown

  NATS v2 has `lame_duck_grace_period` and `lame_duck_duration` which are 10s and 2 minutes respectively by default.  This would mean that over a duration of 2 minutes the client connections (gorouters) would be slowly reconnecting to another NATS server in the cluster.  This would mitigate some of the spikes that happen currently during the reconnection (thundering herd).

  For this to happen, BOSH would have to be changed to send a USR2 signal to the NATS Server before shutting down a node.

Also, for clusters with gorouters in the thousands I'm thinking something also that could be tried is to bump the reading from the current default of `write_deadline = 2s` to about 5s or 10s.  In next release of NATS v2 we will change the default of `write_deadline` to be `10s` as well.

Happy to chat on how to help to move this forward, seems that big CF deployments would benefit a lot from an upgrade :)

**Comment by ameowlia** [comment]
Hi @wallyqs, 

Thank you so much for this comment. Not only do they sound like great suggestions I appreciate you explaining how they would help.

I swear this is still on our radar and is something that we intend to do soon 😅

**Comment by ameowlia** [comment]
@plowin from SAP was asking if we have any OSS information about our intent to do this work. 

❓ @acrmp - can you provide an update here on our current thinking?

**Comment by acrmp** [comment]
We are still keen to upgrade to NATS v2 to ensure we are using an up-to-date version of NATS and for other reasons called out by @wallyqs.

This is a drop-in replacement from the client perspective but as the NATS cluster protocols used by v1 and v2 are incompatible we need to be careful how we approach the upgrade. In particular we are concerned about the potential for partitioning as we roll nats nodes. For Cloud Foundry there is some potential for this to result in application downtime.

tldr; We're excited to get to NATS v2 for the improvements it offers, we're currently in the process of thinking through the implications of the upgrade so we can make the change safely.

**Comment by ameowlia** [comment]
I wanted to give an update about the state of upgrading to nats2.

### Why we want to update
* We should stay on latest releases of dependancies
* Users have been experiencing [issues](https://github.com/cloudfoundry/nats-release/issues/25) with NATS where something causes lots of NATS' clients to reconnect (something this has been a network blip). This then causes NATS' CPU to spike. This then prevents clients from connecting. The clients continue hammering NATS to try to reconnect.  Ad infinitum until NATS is manually restarted.
* We believe that updating will fix this issue, but are not 100% sure. 
* From @pivotal-Josh-Gainey who has does the deepest dive into this issue and reproduction: "Our reproductions were from tcp write timeouts, though the capture from customer appeared to be tcp read timeouts. Unsure how much of a difference this makes but wanted to add because we have not been able to 100% confirm that gnatsd 2.0 resolves the exact issue that the user is experiencing but we have confirmed it resolves our local reproductions. In both scenarios it seems the server mutex lock being held up is what causes existing connections to be serviced but new ones to fail."

### Why upgrading isn't easy
* NATs v1 cannot talk to NATs v2.
* During a rolling deploy, if there are more than 2 nats VMs, then there will be a split brain.
* A short lived split brain situation would cause a blip of route instability. Here are some things that could happen...
  * only some gorouters won't get new routes or will delete routes
  * only some service discovery controllers will get new internal routes or delete internal routes
* During a long lived split brain situation (anything longer than 120s) here are some things that could happen...
  * all of the above
  * anything that has a TTL will be pruned. This includes: internal routes and some component routes (blobstore, cf-mysql-proxy, routing-api).


### Working Status
* Have done exploration work
* Have not decided on path forward
* Have not started execusion
* On horizon to pick up this work "soon"

**Comment by ameowlia** [comment]
We did this! It was released in [v46](https://github.com/cloudfoundry/nats-release/releases/tag/v46)!

**Comment by wallyqs** [comment]
@ameowlia that’s awesome!! thanks for the update!
